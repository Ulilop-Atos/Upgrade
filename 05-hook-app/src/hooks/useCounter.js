import { useState } from 'react';

export const useCounter = ( initialState = 0 ) => {

  const [state, setState] = useState(initialState);
  
  const increment = ( n = 1 ) => {
    setState(state + n );
  };
  
  const decrement = ( n = 1 ) => {
    setState(state - n );
  };
 
  const reset = ( n = initialState ) => {
    setState( initialState );
  };

  return ({
    state,
    decrement,
    increment,
    reset,
  });
}
