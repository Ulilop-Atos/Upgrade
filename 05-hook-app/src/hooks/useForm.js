import PropTypes from 'prop-types';
import { useState } from 'react';

export const useForm = ( initialState = {} ) => {

  const [values, setValues] = useState(initialState);

  const handleInputChange = ({ target }) => {
    setValues({
      ...values,
      [ target.name ]: target.value, 
    });
  }

  return [values, handleInputChange];

}

useForm.propTypes = {
  initialState: PropTypes.object,
}
