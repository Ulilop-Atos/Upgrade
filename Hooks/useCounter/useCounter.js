import { useState } from 'react';

export const useCounter = ( initialCounter = 0 ) => {

  const [counter, setCounter] = useState(initialCounter);
  
  const increment = ( n = 1 ) => {
    if (typeof n !== typeof counter) {
      throw new Error(`Error on increment, ${n} is not a number.`);
    }
    setCounter(counter + n);
  };
  
  const decrement = ( n = 1 ) => {
    if (typeof n !== typeof counter) {
      throw new Error(`Error on decrement, ${n} is not a number.`);
    }    
    setCounter(counter - n);
  };
 
  const reset = ( n = initialCounter ) => {
    if (typeof n !== typeof counter) {
      throw new Error(`Error on reset, ${n} is not a number.`);
    } 
    setCounter( n );
  };

  return ({
    counter,
    decrement,
    increment,
    reset,
  });
}
