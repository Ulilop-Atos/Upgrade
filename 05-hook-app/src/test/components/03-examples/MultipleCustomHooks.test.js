import { shallow } from 'enzyme';
import { MultipleCustomHooks } from '../../../components/03-examples/MultipleCustomHooks';
import { useCounter } from '../../../hooks/useCounter';
import { useFetch } from '../../../hooks/useFetch';
jest.mock('../../../hooks/useFetch');
jest.mock('../../../hooks/useCounter');

describe('Tests for <MultipleCustomHooks />', () => { 
  
  test('should display properly before loading', () => {
    useCounter.mockReturnValue({
      state: 1,
      increment: () => {},
    });
    useFetch.mockReturnValue({
      data:null,
      loading:true,
      error:null,
    });
    const wrapper =  shallow(<MultipleCustomHooks />);
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('.alert').exists()).toBe(true);
   });

  test('should display properly after loading', () => { 
    const quote = 'This is a test quote.';
    const author = 'Test Author';
    useCounter.mockReturnValue({
      state: 1,
      increment: () => {},
    });
    useFetch.mockReturnValue({
      data:[{
        author,
        quote,
      }],
      loading:false,
      error:null,
    });
    const wrapper =  shallow(<MultipleCustomHooks />);
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('.alert').exists()).toBe(false);
    expect(wrapper.find('.mb-0').text().trim()).toBe(quote);
    expect(wrapper.find('.blockquote-footer').text().trim()).toBe(author);
  });

 });