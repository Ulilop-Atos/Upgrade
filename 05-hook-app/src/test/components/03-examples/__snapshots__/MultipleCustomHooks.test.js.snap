// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for <MultipleCustomHooks /> should display properly after loading 1`] = `
<div>
  <h1>
    BreakingBad Quotes
  </h1>
  <hr />
  <blockquote
    className="blockquote text-right"
  >
    <p
      className="mb-0"
    >
       
      This is a test quote.
       
    </p>
    <br />
    <footer
      className="blockquote-footer"
    >
       
      Test Author
       
    </footer>
  </blockquote>
  <button
    className="btn btn-primary"
    onClick={[Function]}
  >
    Next quote
  </button>
</div>
`;

exports[`Tests for <MultipleCustomHooks /> should display properly before loading 1`] = `
<div>
  <h1>
    BreakingBad Quotes
  </h1>
  <hr />
  <div
    className="alert alert-info text-center"
  >
    Loading...
  </div>
  <button
    className="btn btn-primary"
    onClick={[Function]}
  >
    Next quote
  </button>
</div>
`;
